UdpTransport.cs

```diff
@@ -178,24 +241,80 @@ public byte[] Request(IPAddress peer, int port, byte[] buffer, int bufferLength,
 				}
 				if (recv > 0)
 				{
+                    //it is possible that a trap was received and the expected response will be received after this ...
+                    try
+                    {
+                        bool bIsTrap = false;
+
+                        // Check protocol version int 
+                        int ver = SnmpPacket.GetProtocolVersion(inbuffer, recv);
+                        if (ver == (int)SnmpVersion.Ver1)
+                        {
+                            // Parse SNMP Version 1 TRAP packet 
+                            SnmpV1TrapPacket pkt = new SnmpV1TrapPacket();
+                            pkt.decode(inbuffer, recv);
+                            
+                            //if we are here, the packet is a trap
+                            bIsTrap = true;
+                        }
+                        else if(ver == (int)SnmpVersion.Ver2)
+                        {
+                            // Parse SNMP Version 2 TRAP packet 
+                            SnmpV2Packet pkt = new SnmpV2Packet();
+                            pkt.decode(inbuffer, recv);
+
+                            if (pkt.Pdu.Type == PduType.V2Trap) 
+                            {
+                                //if we are here it is a V2Packet
+                                bIsTrap = true;
+                            }
+                        }
+                        else if (ver == (int)SnmpVersion.Ver3)
+                        {
+                            //TODO - evalute if the response is a v3 trap 
+
+                        }
+
+                        if (bIsTrap)
+                        {
+                            //we received a trap, don't use this as a answer!
+
+                            //let's try to wait for another response without sending the request again
+                            retryWithoutSend = true;
+
+                            //continue to receive
+                            continue;
+                        }
+                    }
+                    catch (Exception)
+                    {
+                        //it is no trap!
+                        retryWithoutSend = false;
+                    }
+
+                    //it is no trap!
+                    retryWithoutSend = false;
+
+
```

Comment:
I know it is not so nice in regards of layers to evaluate the application message within the transport protocol. But there was the problem that the response of a SNMP request was not detected correctly in case a trap has been received at the same time.

Reply:
I will never accept this.
and why trap will recevie, trap is receive on another port